GNUmakefile

```diff
@@ -10,7 +12,7 @@ WARNINGS = -Wall -Wextra -Wpedantic -Wformat=2 -Wstrict-aliasing=3 -Wstrict-over
 	-Wfloat-equal -Wcast-align -Wpointer-arith -Wchar-subscripts -Warray-bounds=2 -Wno-unknown-warning-option
 
 override CFLAGS ?= -g -O2 $(WARNINGS)
-override CFLAGS += -std=c99
+override CFLAGS += -std=c99 -fPIC
```

Comment:
Was this needed? The .so's are already built with `-fPIC`

Reply:
@Cloudef  cdl.a isn't currently built with fPIC, so attempting to build bemenu results in 

`/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../x86_64-pc-linux-gnu/bin/ld: cdl.a: relocation R_X86_64_32 against '.rodata' can not be used when making a shared object; recompile with -fPIC`

Reply:
@ambasta does this help? weird that it's not required on linux or osx
```
diff --git a/GNUmakefile b/GNUmakefile
index 70f8577..55751db 100644
--- a/GNUmakefile
+++ b/GNUmakefile
@@ -56,6 +56,7 @@ $(renderers): %: VERSION .git/index | $(libs)
 $(bins): %: | $(libs)
 	$(LINK.c) $(filter %.c %.a,$^) $(LDLIBS) -L. -lbemenu -o $@
 
+cdl.a: private override LDFLAGS += -fPIC
 cdl.a: lib/3rdparty/cdl.c lib/3rdparty/cdl.h
 
 libbemenu.so: private override LDLIBS += -ldl
```

Reply:
It partially helps, w/o the the patch the build log is:

```
$ make clients curses wayland PREFIX=/usr libdir=/lib64
cc -g -O2 -Wall -Wextra -Wpedantic -Wformat=2 -Wstrict-aliasing=3 -Wstrict-overflow=5 -Wstack-usage=12500 -Wfloat-equal -Wcast-align -Wpointer-arith -Wchar-subscripts -Warray-bounds=2 -Wno-unknown-warning-option  -std=c99 -D_FORTIFY_SOURCE=2 -DBM_VERSION=\"0.6.1\" -DBM_PLUGIN_VERSION=\"0.6.1-nogit\" -DINSTALL_LIBDIR=\"/usr/lib64\" -D_DEFAULT_SOURCE -Ilib   -c lib/3rdparty/cdl.c  -o cdl.a
cc -g -O2 -Wall -Wextra -Wpedantic -Wformat=2 -Wstrict-aliasing=3 -Wstrict-overflow=5 -Wstack-usage=12500 -Wfloat-equal -Wcast-align -Wpointer-arith -Wchar-subscripts -Warray-bounds=2 -Wno-unknown-warning-option  -std=c99 -D_FORTIFY_SOURCE=2 -DBM_VERSION=\"0.6.1\" -DBM_PLUGIN_VERSION=\"0.6.1-nogit\" -DINSTALL_LIBDIR=\"/usr/lib64\" -D_DEFAULT_SOURCE -Ilib   -shared -fPIC lib/filter.c lib/item.c lib/library.c lib/list.c lib/menu.c lib/util.c cdl.a -ldl -o libbemenu.so.0.6.1 -Wl,-soname,libbemenu.so.0
/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../x86_64-pc-linux-gnu/bin/ld: cdl.a: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
collect2: error: ld returned 1 exit status
make: *** [GNUmakefile:43: libbemenu.so] Error 1
```

With changes applied it progresses much further
```
$ make clients curses wayland PREFIX=/usr libdir=/lib64
cc -g -O2 -Wall -Wextra -Wpedantic -Wformat=2 -Wstrict-aliasing=3 -Wstrict-overflow=5 -Wstack-usage=12500 -Wfloat-equal -Wcast-align -Wpointer-arith -Wchar-subscripts -Warray-bounds=2 -Wno-unknown-warning-option  -std=c99 -D_FORTIFY_SOURCE=2 -DBM_VERSION=\"0.6.1\" -DBM_PLUGIN_VERSION=\"0.6.1-nogit\" -DINSTALL_LIBDIR=\"/usr/lib64\" -D_DEFAULT_SOURCE -Ilib -fPIC  -c lib/3rdparty/cdl.c  -o cdl.a
cc -g -O2 -Wall -Wextra -Wpedantic -Wformat=2 -Wstrict-aliasing=3 -Wstrict-overflow=5 -Wstack-usage=12500 -Wfloat-equal -Wcast-align -Wpointer-arith -Wchar-subscripts -Warray-bounds=2 -Wno-unknown-warning-option  -std=c99 -D_FORTIFY_SOURCE=2 -DBM_VERSION=\"0.6.1\" -DBM_PLUGIN_VERSION=\"0.6.1-nogit\" -DINSTALL_LIBDIR=\"/usr/lib64\" -D_DEFAULT_SOURCE -Ilib   -shared -fPIC lib/filter.c lib/item.c lib/library.c lib/list.c lib/menu.c lib/util.c cdl.a -ldl -o libbemenu.so.0.6.1 -Wl,-soname,libbemenu.so.0
ln -fs libbemenu.so.0.6.1 libbemenu.so.0
ln -fs libbemenu.so.0.6.1 libbemenu.so
cc -g -O2 -Wall -Wextra -Wpedantic -Wformat=2 -Wstrict-aliasing=3 -Wstrict-overflow=5 -Wstack-usage=12500 -Wfloat-equal -Wcast-align -Wpointer-arith -Wchar-subscripts -Warray-bounds=2 -Wno-unknown-warning-option  -std=c99 -D_FORTIFY_SOURCE=2 -DBM_VERSION=\"0.6.1\" -DBM_PLUGIN_VERSION=\"0.6.1-nogit\" -DINSTALL_LIBDIR=\"/usr/lib64\" -D_DEFAULT_SOURCE -Ilib   -c client/common/common.c  -o common.a
cc -g -O2 -Wall -Wextra -Wpedantic -Wformat=2 -Wstrict-aliasing=3 -Wstrict-overflow=5 -Wstack-usage=12500 -Wfloat-equal -Wcast-align -Wpointer-arith -Wchar-subscripts -Warray-bounds=2 -Wno-unknown-warning-option  -std=c99 -D_FORTIFY_SOURCE=2 -DBM_VERSION=\"0.6.1\" -DBM_PLUGIN_VERSION=\"0.6.1-nogit\" -DINSTALL_LIBDIR=\"/usr/lib64\" -D_DEFAULT_SOURCE -Ilib   common.a client/bemenu.c  -L. -lbemenu -o bemenu
cc -g -O2 -Wall -Wextra -Wpedantic -Wformat=2 -Wstrict-aliasing=3 -Wstrict-overflow=5 -Wstack-usage=12500 -Wfloat-equal -Wcast-align -Wpointer-arith -Wchar-subscripts -Warray-bounds=2 -Wno-unknown-warning-option  -std=c99 -D_FORTIFY_SOURCE=2 -DBM_VERSION=\"0.6.1\" -DBM_PLUGIN_VERSION=\"0.6.1-nogit\" -DINSTALL_LIBDIR=\"/usr/lib64\" -D_DEFAULT_SOURCE -Ilib   common.a client/bemenu-run.c  -L. -lbemenu -o bemenu-run
cc -g -O2 -Wall -Wextra -Wpedantic -Wformat=2 -Wstrict-aliasing=3 -Wstrict-overflow=5 -Wstack-usage=12500 -Wfloat-equal -Wcast-align -Wpointer-arith -Wchar-subscripts -Warray-bounds=2 -Wno-unknown-warning-option  -std=c99 -D_FORTIFY_SOURCE=2 -DBM_VERSION=\"0.6.1\" -DBM_PLUGIN_VERSION=\"0.6.1-nogit\" -DINSTALL_LIBDIR=\"/usr/lib64\" -D_DEFAULT_SOURCE -Ilib -I/usr/include/ncursesw    -shared -fPIC lib/renderers/curses/curses.c -lncursesw -ltinfow  -lm -L. -lbemenu -o bemenu-renderer-curses.so
wayland-scanner private-code < "$(pkg-config --variable=pkgdatadir wayland-protocols)/stable/xdg-shell/xdg-shell.xml" > lib/renderers/wayland/xdg-shell.c
cc -g -O2 -Wall -Wextra -Wpedantic -Wformat=2 -Wstrict-aliasing=3 -Wstrict-overflow=5 -Wstack-usage=12500 -Wfloat-equal -Wcast-align -Wpointer-arith -Wchar-subscripts -Warray-bounds=2 -Wno-unknown-warning-option  -std=c99 -D_FORTIFY_SOURCE=2 -DBM_VERSION=\"0.6.1\" -DBM_PLUGIN_VERSION=\"0.6.1-nogit\" -DINSTALL_LIBDIR=\"/usr/lib64\" -D_DEFAULT_SOURCE -Ilib -I/usr/lib64/libffi/include    -c lib/renderers/wayland/xdg-shell.c  -o xdg-shell.a
wayland-scanner private-code < lib/renderers/wayland/wlr-layer-shell-unstable-v1.xml > lib/renderers/wayland/wlr-layer-shell-unstable-v1.c
wayland-scanner client-header < lib/renderers/wayland/wlr-layer-shell-unstable-v1.xml > lib/renderers/wayland/wlr-layer-shell-unstable-v1.h
cc -g -O2 -Wall -Wextra -Wpedantic -Wformat=2 -Wstrict-aliasing=3 -Wstrict-overflow=5 -Wstack-usage=12500 -Wfloat-equal -Wcast-align -Wpointer-arith -Wchar-subscripts -Warray-bounds=2 -Wno-unknown-warning-option  -std=c99 -D_FORTIFY_SOURCE=2 -DBM_VERSION=\"0.6.1\" -DBM_PLUGIN_VERSION=\"0.6.1-nogit\" -DINSTALL_LIBDIR=\"/usr/lib64\" -D_DEFAULT_SOURCE -Ilib -I/usr/lib64/libffi/include    -c lib/renderers/wayland/wlr-layer-shell-unstable-v1.c  -o wlr-layer-shell.a
cc -g -O2 -Wall -Wextra -Wpedantic -Wformat=2 -Wstrict-aliasing=3 -Wstrict-overflow=5 -Wstack-usage=12500 -Wfloat-equal -Wcast-align -Wpointer-arith -Wchar-subscripts -Warray-bounds=2 -Wno-unknown-warning-option  -std=c99 -D_FORTIFY_SOURCE=2 -DBM_VERSION=\"0.6.1\" -DBM_PLUGIN_VERSION=\"0.6.1-nogit\" -DINSTALL_LIBDIR=\"/usr/lib64\" -D_DEFAULT_SOURCE -Ilib -I/usr/lib64/libffi/include    -c lib/renderers/wayland/xdg-output-unstable-v1.c  -o xdg-output.a
cc -g -O2 -Wall -Wextra -Wpedantic -Wformat=2 -Wstrict-aliasing=3 -Wstrict-overflow=5 -Wstack-usage=12500 -Wfloat-equal -Wcast-align -Wpointer-arith -Wchar-subscripts -Warray-bounds=2 -Wno-unknown-warning-option  -std=c99 -D_FORTIFY_SOURCE=2 -DBM_VERSION=\"0.6.1\" -DBM_PLUGIN_VERSION=\"0.6.1-nogit\" -DINSTALL_LIBDIR=\"/usr/lib64\" -D_DEFAULT_SOURCE -Ilib -I/usr/lib64/libffi/include -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/freetype2 -I/usr/include/uuid -I/usr/include/libpng16 -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/fribidi    -shared -fPIC lib/renderers/wayland/wayland.c lib/renderers/wayland/registry.c lib/renderers/wayland/window.c xdg-shell.a wlr-layer-shell.a xdg-output.a -lwayland-client -lpangocairo-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lharfbuzz -lcairo -lxkbcommon  -L. -lbemenu -o bemenu-renderer-wayland.so
/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../x86_64-pc-linux-gnu/bin/ld: xdg-shell.a: warning: relocation in read-only section `.rodata'
/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../x86_64-pc-linux-gnu/bin/ld: warning: creating DT_TEXTREL in a shared object
```

Do note that this occurs on linux for me (Gentoo)

Reply:
I think this occurs because my gcc is compiled w/o pie support. A global fPIC is a solution to that

Reply:
That could explain it:
```
diff --git a/GNUmakefile b/GNUmakefile
index 70f8577..568cbb9 100644
--- a/GNUmakefile
+++ b/GNUmakefile
@@ -56,6 +56,7 @@ $(renderers): %: VERSION .git/index | $(libs)
 $(bins): %: | $(libs)
 	$(LINK.c) $(filter %.c %.a,$^) $(LDLIBS) -L. -lbemenu -o $@
 
+cdl.a: private override LDFLAGS += -fPIC
 cdl.a: lib/3rdparty/cdl.c lib/3rdparty/cdl.h
 
 libbemenu.so: private override LDLIBS += -ldl
@@ -84,10 +85,13 @@ lib/renderers/wayland/xdg-output-unstable-v1.h: lib/renderers/wayland/xdg-output
 lib/renderers/wayland/xdg-output-unstable-v1.c: lib/renderers/wayland/xdg-output-unstable-v1.xml
 	wayland-scanner private-code < $^ > $@
 
+xdg-shell.a: private override LDFLAGS += -fPIC
 xdg-shell.a: private override CPPFLAGS += $(shell pkg-config --cflags-only-I wayland-client)
 xdg-shell.a: lib/renderers/wayland/xdg-shell.c
+wlr-layer-shell.a: private override LDFLAGS += -fPIC
 wlr-layer-shell.a: private override CPPFLAGS += $(shell pkg-config --cflags-only-I wayland-client)
 wlr-layer-shell.a: lib/renderers/wayland/wlr-layer-shell-unstable-v1.c lib/renderers/wayland/wlr-layer-shell-unstable-v1.h
+xdg-output.a: private override LDFLAGS += -fPIC
 xdg-output.a: private override CPPFLAGS += $(shell pkg-config --cflags-only-I wayland-client)
 xdg-output.a: lib/renderers/wayland/xdg-output-unstable-v1.c lib/renderers/wayland/xdg-output-unstable-v1.h
 bemenu-renderer-wayland.so: private override LDLIBS += $(shell pkg-config --libs wayland-client cairo pango pangocairo xkbcommon)
```

Does this work? If static lib is linked against shared lib it needs -fPIC but if linked against binary it should not be needed. That's why I did not add it to the common .a rule.

Reply:
Yes, the latest patch can be verfied working for me.

Reply:
I'll push it and make a new release